### PR TITLE
refactor: delete all trackers when finalizing CCTX

### PR DESCRIPTION
**Pull request by kingpinXD**
# Description

Closes https://github.com/zeta-chain/node/issues/2610
- Try to delete all trackers associated with the cctx when finalizing it. If the tracker doesn't exist, nothing happens
- Try to update the pending nonce for all nonces associated with the tracker, if the nonce has been previously updated, nothing happens 
- Fixes setting the ballot index properly for cases when a revert is added . (previously the ballot index was getting set to the revert tx instead of the original outbound )

# How Has This Been Tested?

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [ ] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions

## Summary by CodeRabbit

- **New Features**
  - Added a validation check for outbound transaction trackers in the local end-to-end testing framework.
  - Introduced a new method to set the outbound ballot index within the CrossChainTx structure.
  
- **Bug Fixes**
  - Improved handling of outbound transaction states by simplifying method signatures and enhancing functionality.

- **Documentation**
  - Minor formatting adjustments in command-line documentation for clarity.

- **Tests**
  - Expanded test coverage for outbound transaction handling and nonce removal functionality.

### CODE REVIEW

File: x/crosschain/types/cctx_test.go [hunk 1]
```diff
@@ -16,6 +16,12 @@ import (
 	observertypes "github.com/zeta-chain/zetacore/x/observer/types"
 )
 
+func TestCrossChainTx_SetOutboundBallot(t *testing.T) {
+	cctx := sample.CrossChainTx(t, "test")
+	ballotIndex := sample.ZetaIndex(t)
+	cctx.SetOutboundBallotIndex(ballotIndex)
+	require.Equal(t, ballotIndex, cctx.GetCurrentOutboundParam().BallotIndex)
+}
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply x/crosschain/types/cctx_test.go:24]: That's a good suggestion

File: cmd/zetae2e/local/local.go [hunk 2]
```diff
@@ -362,10 +362,18 @@ func localE2ETest(cmd *cobra.Command, _ []string) {
...
 
 	os.Exit(0)
 }
 
+func checkTrackers(ctx context.Context, deployRunner *runner.E2ERunner) {
+	// get all trackers
+	res, err := deployRunner.CctxClient.OutTxTrackerAll(ctx, &crosschaintypes.QueryAllOutboundTrackerRequest{})
+	require.NoError(deployRunner, err)
+	require.Empty(deployRunner, res.OutboundTracker, "there should be no trackers")
+}
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply cmd/zetae2e/local/local.go:375]: This check is not needed as this is test, and this call should not return an error

File: x/crosschain/keeper/msg_server_vote_outbound_tx_test.go [hunk 3]
```diff
@@ -465,7 +469,7 @@ func TestKeeper_SaveFailedOutbound(t *testing.T) {
 			HashList: nil,
 		})
 		cctx.CctxStatus.Status = types.CctxStatus_PendingOutbound
-		k.SaveFailedOutbound(ctx, cctx, sample.String(), sample.ZetaIndex(t))
+		k.SaveFailedOutbound(ctx, cctx, sample.String())
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply x/crosschain/keeper/msg_server_vote_outbound_tx_test.go:472]: This suggestion is wrong , the string argument is required for the error message

**Review by ws4charlie — APPROVED** [review]

**Review by swift1337 — APPROVED** [review]

File: cmd/zetae2e/local/local.go [hunk 4]
```diff
@@ -366,6 +366,16 @@ func localE2ETest(cmd *cobra.Command, _ []string) {
 	os.Exit(0)
 }
 
+func checkTrackers(deployRunner *runner.E2ERunner) {
```
Comment (lumtis) [inline cmd/zetae2e/local/local.go:369]: Could be implemented in `e2e` package instead

File: x/observer/keeper/pending_nonces_test.go [hunk 5]
```diff
@@ -136,4 +137,54 @@ func TestKeeper_RemoveFromPendingNonces(t *testing.T) {
...
 		require.True(t, nonceUpdated)
 	})
+
+	t.Run("test removal within range only using fixed value", func(t *testing.T) {
+		k, ctx, _, _ := keepertest.ObserverKeeper(t)
+
+		tss := sample.Tss()
+		// make nonces and pubkey deterministic for test
+		chainIDS := []int64{chains.GoerliLocalnet.ChainId, chains.BitcoinTestnet.ChainId, chains.BscTestnet.ChainId}
+		pn := make([]types.PendingNonces, len(chainIDS))
```
Comment (lumtis) [inline x/observer/keeper/pending_nonces_test.go:147]:
```suggestion
		pendingNonces := make([]types.PendingNonces, len(chainIDS))
```
better to keep explicit variable names

File: x/observer/keeper/pending_nonces_test.go [hunk 6]
```diff
@@ -136,4 +137,54 @@ func TestKeeper_RemoveFromPendingNonces(t *testing.T) {
...
+				Tss:       tss.TssPubkey,
+			}
+		}
+		for _, pendingNonce := range pn {
+			k.SetPendingNonces(ctx, pendingNonce)
+		}
+
+		// remove from pending nonces
+		k.RemoveFromPendingNonces(ctx, tss.TssPubkey, chains.GoerliLocalnet.ChainId, 1)
+		pnGoerli, found := k.GetPendingNonces(ctx, tss.TssPubkey, chains.GoerliLocalnet.ChainId)
```
Comment (lumtis) [inline x/observer/keeper/pending_nonces_test.go:163]:
```suggestion
		actualPendingNonces, found := k.GetPendingNonces(ctx, tss.TssPubkey, chains.GoerliLocalnet.ChainId)
```

File: x/crosschain/types/cctx.go [hunk 7]
```diff
@@ -164,6 +164,10 @@ func (m CrossChainTx) GetCCTXIndexBytes() ([32]byte, error) {
 	return sendHash, nil
 }
 
+func (m CrossChainTx) SetOutboundBallotIndex(index string) {
```
Comment (lumtis) [inline x/crosschain/types/cctx.go:167]: Let's add comment, it adds the index specifically to the last outbound

**Review by lumtis — COMMENTED** [review]
>Fixes setting the ballot index properly for cases when a revert is added . (previously the ballot index was getting set to the revert tx instead of the original outbound )

What is the change reflecting this?

**Comment by kingpinXD** [comment]
> > Fixes setting the ballot index properly for cases when a revert is added . (previously the ballot index was getting set to the revert tx instead of the original outbound )
> 
> What is the change reflecting this?

https://github.com/zeta-chain/zeta-node/blob/22f14acd3f5e32aeb370866cb6f3ab2576865fe5/x/crosschain/keeper/msg_server_vote_outbound_tx.go#L93-L94

Earlier this ballot was set in the SaveOutbound. 
In case if a revert was added , the ballot index for outboundParam[0] was in fact added to outboundParam[1]

**Review by lumtis — APPROVED** [review]
LGTM
@kingpinXD please use explicit PR title

**Comment by kingpinXD** [comment]
> LGTM @kingpinXD please use explicit PR title

@lumtis could you give me an example of what you mean, Imo the title reflects the change 
Are you referring to the fact that pending nonces are also cleaned ? ( I did not add that in as the change keeps the same behaviour as before )

**Comment by lumtis** [comment]
> > LGTM @kingpinXD please use explicit PR title
> 
> @lumtis could you give me an example of what you mean, Imo the title reflects the change Are you referring to the fact that pending nonces are also cleaned ? ( I did not add that in as the change keeps the same behaviour as before )

I changed the name from `tracker cleanup` to `delete all trackers when finalizing CCTX` before putting this comment.
 `tracker cleanup` doesn't allow to understand the change by looking at PR title.

**Comment by kingpinXD** [comment]
> > > LGTM @kingpinXD please use explicit PR title
> > 
> > 
> > @lumtis could you give me an example of what you mean, Imo the title reflects the change Are you referring to the fact that pending nonces are also cleaned ? ( I did not add that in as the change keeps the same behaviour as before )
> 
> I changed the name from `tracker cleanup` to `delete all trackers when finalizing CCTX` before putting this comment. `tracker cleanup` doesn't allow to understand the change by looking at PR title.

okay that makes sense , thanks for the edit . 
Got confused that you were requesting a further change

**Pull request merged**